lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,959 @@
-/// Return argument attributes with annotations removed.
-void GrandCentralVisitor::handlePorts(Operation *op) {
-
-  SmallVector<Attribute> newArgAttrs;
-  auto ports = cast<FModuleLike>(op).getPorts();
-  for (size_t i = 0, e = ports.size(); i != e; ++i) {
-    auto port = ports[i];
-    handleRefLike(op, port.annotations, port.type);
-    newArgAttrs.push_back(port.annotations.getArrayAttr());
-  }
-
-  op->setAttr("portAnnotations", ArrayAttr::get(op->getContext(), newArgAttrs));
-}
+/// Return the module that is associated with this value.
+///
+/// TODO: Refactor this to not use `getReferencedModule`.
+FModuleLike GrandCentralPass::getEnclosingModule(Value value) {
+  if (auto blockArg = value.dyn_cast<BlockArgument>())
+    return cast<FModuleOp>(blockArg.getOwner()->getParentOp());
 
-void GrandCentralVisitor::visitDecl(InstanceOp op) {
-
-  // If this instance's underlying module has a "companion" annotation, then
-  // move this onto the actual instance op.
-  AnnotationSet annotations(op.getReferencedModule());
-  if (auto anno = annotations.getAnnotation(
-          "sifive.enterprise.grandcentral.ViewAnnotation")) {
-    auto tpe = anno.getAs<StringAttr>("type");
-    if (!tpe) {
-      op.getReferencedModule()->emitOpError(
-          "contains a ViewAnnotation that does not contain a \"type\" field");
-      failed = true;
-      return;
-    }
-    if (tpe.getValue() == "companion")
-      op->setAttr("lowerToBind", BoolAttr::get(op.getContext(), true));
-  }
+  auto *op = value.getDefiningOp();
+  if (InstanceOp instance = dyn_cast<InstanceOp>(op))
+    return instance.getReferencedModule();
 
-  visitUnhandledDecl(op);
+  return op->getParentOfType<FModuleOp>();
 }
 
+/// This method contains the business logic of this pass.
 void GrandCentralPass::runOnOperation() {
+  LLVM_DEBUG(llvm::dbgs() << "===- Running Grand Central Views/Interface Pass "
+                             "-----------------------------===\n");
+
   CircuitOp circuitOp = getOperation();
 
   AnnotationSet annotations(circuitOp);
   if (annotations.empty())
-    return;
+    return markAllAnalysesPreserved();
+
+  // Pull out the extraction info if it exists.
+  bool removalError = false, workToDo = false;
+  annotations.removeAnnotations([&](Annotation anno) {
+    if (anno.isClass(
+            "sifive.enterprise.grandcentral.ExtractGrandCentralAnnotation")) {
+      if (maybeExtractInfo.hasValue()) {
+        emitCircuitError("more than one 'ExtractGrandCentralAnnotation' was "
+                         "found, but exactly one must be provided");
+        removalError = true;
+        return false;
+      }
+
+      maybeExtractInfo = {anno.getMember<StringAttr>("directory").getValue(),
+                          anno.getMember<StringAttr>("filename").getValue()};
+      return false;
+    }
+    workToDo |=
+        anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType");
+    return false;
```

Comment:
Would it make sense to remove these annotations with `return true` while you already have them, and populating a local worklist of `AugmentedBundleType`s?

Reply:
Yes, that is cleaner. Will switch to this in a rebase...